src/lib/pickles/pickles.ml

```diff
@@ -974,16 +975,41 @@ let compile_promise :
            * (max_proofs_verified, max_proofs_verified) Proof.t )
            Promise.t )
          H3_2.T(Prover).t =
+ (* This function is an adapter between the user-facing Pickles.compile API
+    and the underlying Make(_).compile function which builds the circuits.
+ *)
  fun ?self ?(cache = []) ?disk_keys ?(return_early_digest_exception = false)
-     (module A_var) (module A_value) ~public_input ~auxiliary_typ ~branches
-     ~max_proofs_verified ~name ~constraint_constants ~choices ->
+     ~public_input ~auxiliary_typ ~branches ~max_proofs_verified ~name
+     ~constraint_constants ~choices () ->
   let self =
     match self with
     | None ->
         { Tag.id = Type_equal.Id.create ~name sexp_of_opaque; kind = Compiled }
     | Some self ->
         self
   in
+  (* Extract to_fields methods from the public input declaration. *)
+  let (a_var_to_fields : a_var -> _), (a_value_to_fields : a_value -> _) =
+    match public_input with
+    | Input (Typ typ) ->
+        ( (fun x -> fst (typ.var_to_fields x))
+        , fun x -> fst (typ.value_to_fields x) )
+    | Output _ ->
+        ((fun () -> [||]), fun () -> [||])
+    | Input_and_output (Typ typ, _) ->
+        ( (fun x -> fst (typ.var_to_fields x))
+        , fun x -> fst (typ.value_to_fields x) )
+  in
+  let module A_var = struct
+    type t = a_var
+
+    let to_field_elements = a_var_to_fields
+  end in
+  let module A_value = struct
+    type t = a_value
+
+    let to_field_elements = a_value_to_fields
+  end in
```

Comment:
doesn't this look very different from what you deleted up there?

```
  let to_field_elements (t : t) = [| Data_as_hash.hash t |]
```

Reply:
It does, but it's doing the same thing under the hood (via the `Typ.t`).